cluster/job_test.go

```diff
@@ -11,23 +11,268 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestSchedule(t *testing.T) {
-	t.Parallel()
+func TestMakeWaitForInterval(t *testing.T) {
+	t.Run("panics on invalid interval", func(t *testing.T) {
+		assert.Panics(t, func() {
+			MakeWaitForInterval(0)
+		})
+	})
 
-	makeKey := func() string {
-		return model.NewId()
+	const neverRun = -1 * time.Second
+
+	testCases := []struct {
+		Description  string
+		Interval     time.Duration
+		LastFinished time.Duration
+		Expected     time.Duration
+	}{
+		{
+			"never run, 5 minutes",
+			5 * time.Minute,
+			neverRun,
+			0,
+		},
+		{
+			"run 1 minute ago, 5 minutes",
+			5 * time.Minute,
+			-1 * time.Minute,
+			4 * time.Minute,
+		},
+		{
+			"run 2 minutes ago, 5 minutes",
+			5 * time.Minute,
+			-2 * time.Minute,
+			3 * time.Minute,
+		},
+		{
+			"run 4 minutes 30 seconds ago, 5 minutes",
+			5 * time.Minute,
+			-4*time.Minute - 30*time.Second,
+			30 * time.Second,
+		},
+		{
+			"run 4 minutes 59 seconds ago, 5 minutes",
+			5 * time.Minute,
+			-4*time.Minute - 59*time.Second,
+			1 * time.Second,
+		},
+		{
+			"never run, 1 hour",
+			1 * time.Hour,
+			neverRun,
+			0,
+		},
+		{
+			"run 1 minute ago, 1 hour",
+			1 * time.Hour,
+			-1 * time.Minute,
+			59 * time.Minute,
+		},
+		{
+			"run 20 minutes ago, 1 hour",
+			1 * time.Hour,
+			-20 * time.Minute,
+			40 * time.Minute,
+		},
+		{
+			"run 55 minutes 30 seconds ago, 1 hour",
+			1 * time.Hour,
+			-55*time.Minute - 30*time.Second,
+			4*time.Minute + 30*time.Second,
+		},
+		{
+			"run 59 minutes 59 seconds ago, 1 hour",
+			1 * time.Hour,
+			-59*time.Minute - 59*time.Second,
+			1 * time.Second,
+		},
 	}
 
-	t.Run("invalid interval", func(t *testing.T) {
-		t.Parallel()
+	for _, testCase := range testCases {
+		t.Run(testCase.Description, func(t *testing.T) {
+			now := time.Now()
 
-		mockPluginAPI := newMockPluginAPI(t)
+			var lastFinished time.Time
+			if testCase.LastFinished != neverRun {
+				lastFinished = now.Add(testCase.LastFinished)
+			}
+
+			actual := MakeWaitForInterval(testCase.Interval)(now, jobMetadata{
```

Comment:
Comment for off-topic discussion - This function takes the value of `Now` which makes it safe to use as a point of reference for testing. I'm not sure if the `cluster` package ever uses `time.Now()` in its source, but I thought it may be beneficial to mention the package https://github.com/benbjohnson/clock, used to manipulate the "system's" `Now` time during testing.

I've made a function `var nowFunc = time.Now` in my source, then used the `clock` package in the test, temporarily assigning the `nowFunc` to the clocks `Now` function. Depending on the amount of time-based functionality in the `cluster` package, it may be beneficial to introduce something like `clock` in the tests. I'm curious what your thoughts are on this @lieut-data

Reply:
Ooh, that's a very slick approach. It's always a pain mocking the value of `time.Now()`, since it never feels like a factorization that improves the code: just something awkward that's necessary. I'll make a point of (asynchronously) revisiting the clock package you mention and seeing if things can be simpler as a result. Thank you!!

Reply:
0/5 I had the exact same problem in Solar Lottery and decided to stick with the less slick, intention-revealing `now` parameter. I'd be ok going either way, but we should try to be consistent.

Reply:
I thought about this, and using the `now` parameter makes much more sense that assigning a `nowFunc` at package level. Using the `clock` package in the test, while using the `now` function parameter, gives the test a flexible way to provide the dependency.